### PR TITLE
Unit test for QueryUtils, application, plugin

### PR DESCRIPTION
**Pull request by KishoreKicha14**
#### **Description**
- Implemented unit tests for the following components:
  - `QueryUtils.ts`
  - `application.tsx`
  - `plugin.ts`
- Improved test coverage across the codebase:
  - **`QueryUtils.ts` coverage increased from `0%` to `85.71%`.**
  - **Achieved `100%` coverage for `application.tsx` and `plugin.ts`.**
  - **Overall Query Insights coverage improved from `82.57%` to `85.71%`.**
- Enhanced test reliability and ensured comprehensive validation.

### Issues Resolved
_List any issues this PR will resolve, e.g. Closes [...]._ 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/utils/QueryUtils.test.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { retrieveQueryById } from './QueryUtils'; // Update with the correct path
```
Comment (ansjcy) [inline common/utils/QueryUtils.test.ts:6]: Could you remove this? 
```
// Update with the correct path
```

File: common/utils/QueryUtils.test.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      if (endpoint.includes('latency')) {
+        return Promise.resolve({
+          response: { top_queries: [{ id: 'query-123', query: 'SELECT * FROM test' }] },
+        });
+      }
+      return Promise.resolve({ response: { top_queries: [] } });
+    });
+
+    const result = await retrieveQueryById(mockCore, dataSourceId, start, end, id);
+    expect(result).toEqual({ id: 'query-123', query: 'SELECT * FROM test' });
```
Comment (ansjcy) [inline common/utils/QueryUtils.test.ts:31]: Why is this SQL query?

File: common/utils/QueryUtils.test.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+import { retrieveQueryById } from './QueryUtils'; // Update with the correct path
+const mockHttpGet = jest.fn();
+const mockCore = { http: { get: mockHttpGet } };
+
+describe('retrieveQueryById', () => {
+  const dataSourceId = 'test-ds';
+  const start = '2025-01-01T00:00:00Z';
+  const end = '2025-01-31T23:59:59Z';
+  const id = 'query-123';
```
Comment (ansjcy) [inline common/utils/QueryUtils.test.ts:14]: Could you use the uuid format for the id?

File: common/utils/QueryUtils.test.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should return the first top query from the fastest responding API', async () => {
```
Comment (ansjcy) [inline common/utils/QueryUtils.test.ts:20]: I don't get this description,`fastest` API might not be accurate.

File: public/application.test.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    );
+
+    expect(typeof unmount).toBe('function');
+  });
+
+  it('should unmount the component when the returned function is called', () => {
+    const unmount = renderApp(coreMock, depsStartMock, paramsMock, dataSourceManagementMock);
+    unmount();
+
+    expect(ReactDOM.unmountComponentAtNode).toHaveBeenCalledWith(paramsMock.element); // ✅ Corrected
```
Comment (ansjcy) [inline public/application.test.tsx:52]: what is this comment? 
```
// ✅ Corrected
```

File: public/plugin.test.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+    expect(renderApp).toHaveBeenCalled();
+  });
+
+  it('should add the navigation link to nav group', () => {
+    plugin.setup(coreSetupMock, {} as any);
+    expect(addNavLinksMock).toHaveBeenCalled();
+  });
+
+  it('should return empty start and stop methods', () => {
```
Comment (ansjcy) [inline public/plugin.test.tsx:72]: why is this test needed?

File: public/application.test.tsx (hunk 5 above)
Comment (KishoreKicha14) [reply public/application.test.tsx]: Removed

File: common/utils/QueryUtils.test.ts (hunk 4 above)
Comment (KishoreKicha14) [reply common/utils/QueryUtils.test.ts]: Changed the description

File: common/utils/QueryUtils.test.ts (hunk 3 above)
Comment (KishoreKicha14) [reply common/utils/QueryUtils.test.ts]: Added uuid

File: common/utils/QueryUtils.test.ts (hunk 2 above)
Comment (KishoreKicha14) [reply common/utils/QueryUtils.test.ts]: Changed to OpenSearch Query DSL

File: common/utils/QueryUtils.test.ts (hunk 1 above)
Comment (KishoreKicha14) [reply common/utils/QueryUtils.test.ts:6]: Removed

File: public/plugin.test.tsx (hunk 6 above)
Comment (KishoreKicha14) [reply public/plugin.test.tsx:72]: Verifies no unnecessary logic in start and stop which prevents future bugs

File: common/utils/QueryUtils.test.ts (hunk 1 above)
Comment (ansjcy) [reply common/utils/QueryUtils.test.ts:6]: I don't see if removed

File: common/utils/QueryUtils.test.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should return the top query', async () => {
+    mockHttpGet.mockImplementation((endpoint) => {
+      if (endpoint.includes('latency')) {
+        return Promise.resolve({
+          response: {
+            top_queries: [
```
Comment (ansjcy) [inline common/utils/QueryUtils.test.ts:25]: Can't we just use the mockQueries in the utils? instead of creating it here.

File: public/application.test.tsx [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+          core={coreMock}
+          depsStart={depsStartMock}
+          params={paramsMock}
+          dataSourceManagement={dataSourceManagementMock}
+        />
+      </Router>,
+      paramsMock.element
+    );
+
+    expect(typeof unmount).toBe('function');
```
Comment (ansjcy) [inline public/application.test.tsx:45]: I don't see any value of this`expect` here, especially when we have the tests on line 48.

File: public/plugin.test.tsx [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+      application: {
+        register: jest.fn(),
+      },
+      chrome: {
+        navGroup: {
+          addNavLinksToGroup: jest.fn(),
+        },
+      },
+      getStartServices: jest.fn().mockResolvedValue([coreStartMock, {}]),
+    } as unknown) as jest.Mocked<CoreSetup>;
```
Comment (ansjcy) [inline public/plugin.test.tsx:33]: Let's not cast to unknown, maybe reuse the coreMock in OSD

File: public/plugin.test.tsx [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+    coreStartMock = {} as jest.Mocked<CoreStart>;
```
Comment (ansjcy) [inline public/plugin.test.tsx:35]: Same - reuse the coreMock in OSD

File: public/plugin.test.tsx [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    addNavLinksMock = coreSetupMock.chrome.navGroup.addNavLinksToGroup;
+  });
+
+  it('should register the application in setup', () => {
+    plugin.setup(coreSetupMock, {} as any);
+
+    expect(registerMock).toHaveBeenCalledWith(
+      expect.objectContaining({
+        id: PLUGIN_NAME,
+        title: 'Query Insights',
```
Comment (ansjcy) [inline public/plugin.test.tsx:48]: There a several other fields being registered, could you include them all?

File: public/plugin.test.tsx [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+    const appRegistration = registerMock.mock.calls[0][0];
+    expect(appRegistration).toBeDefined();
+
+    const paramsMock = { element: document.createElement('div') };
+    const mountFunction = appRegistration.mount;
+
+    await mountFunction(paramsMock);
+
+    expect(renderApp).toHaveBeenCalled();
```
Comment (ansjcy) [inline public/plugin.test.tsx:64]: There's no validation check that renderApp receives valid arguments.

File: public/plugin.test.tsx [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+  });
+
+  it('should add the navigation link to nav group', () => {
+    plugin.setup(coreSetupMock, {} as any);
+    expect(addNavLinksMock).toHaveBeenCalled();
```
Comment (ansjcy) [inline public/plugin.test.tsx:69]: Could you add property checks here as well?

**Review by ansjcy — CHANGES_REQUESTED** [review]

File: common/utils/QueryUtils.test.ts [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    const result = await retrieveQueryById(mockCore, dataSourceId, start, end, id);
+    expect(result).toBeNull();
+  });
+
+  it('should handle API errors gracefully and return null', async () => {
+    mockHttpGet.mockRejectedValue(new Error('API error'));
+
+    const result = await retrieveQueryById(mockCore, dataSourceId, start, end, id);
+    expect(result).toBeNull();
+  });
```
[the remark this answers is not in the document]
Comment (deshsidd) [reply common/utils/QueryUtils.test.ts:66]: @KishoreKicha14 Did we follow-up on this? I don't see the changes here

File: public/plugin.test.tsx (hunk 11 above)
Comment (deshsidd) [reply public/plugin.test.tsx:36]: Are all the fields added here for testing?

**Review by deshsidd — CHANGES_REQUESTED** [review]

File: common/utils/QueryUtils.test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (KishoreKicha14) [reply common/utils/QueryUtils.test.ts:97]: Updated the edge cases

File: public/plugin.test.tsx (hunk 11 above)
Comment (KishoreKicha14) [reply public/plugin.test.tsx:36]: Yes I added all the fields

File: common/utils/QueryUtils.test.ts (hunk 7 above)
Comment (KishoreKicha14) [reply common/utils/QueryUtils.test.ts]: Updated

File: public/application.test.tsx (hunk 8 above)
Comment (KishoreKicha14) [reply public/application.test.tsx]: updated

File: common/utils/QueryUtils.test.ts [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { retrieveQueryById } from './QueryUtils'; // Update with the correct path
+const mockHttpGet = jest.fn();
+const mockCore = { http: { get: mockHttpGet } };
+
+describe('retrieveQueryById', () => {
+  const dataSourceId = 'test-ds';
```
[the remark this answers is not in the document]
Comment (KishoreKicha14) [reply common/utils/QueryUtils.test.ts]: updated

File: common/utils/QueryUtils.test.ts [hunk 16]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { retrieveQueryById } from './QueryUtils'; // Update with the correct path
+const mockHttpGet = jest.fn();
+const mockCore = { http: { get: mockHttpGet } };
+
+describe('retrieveQueryById', () => {
```
[the remark this answers is not in the document]
Comment (KishoreKicha14) [reply common/utils/QueryUtils.test.ts]: updated the  test description

File: common/utils/QueryUtils.test.ts [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+            top_queries: [
+              {
+                id: '1e5fde5b-c85f-419e-b8b6-3f43b3da4d59',
+                query: JSON.stringify({ match: { user_action: 'login_attempt' } }),
+              },
+            ],
+          },
+        });
+      }
+      return Promise.resolve({ response: { top_queries: [] } });
```
[the remark this answers is not in the document]
Comment (KishoreKicha14) [reply common/utils/QueryUtils.test.ts]: asserted with the exact expected endpoint url

File: common/utils/QueryUtils.test.ts [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+      if (endpoint.includes('latency')) {
+        return Promise.resolve({
+          response: {
+            top_queries: [
+              {
+                id: '1e5fde5b-c85f-419e-b8b6-3f43b3da4d59',
+                query: JSON.stringify({ match: { user_action: 'login_attempt' } }),
+              },
+            ],
+          },
```
[the remark this answers is not in the document]
Comment (KishoreKicha14) [reply common/utils/QueryUtils.test.ts]: Added

File: public/plugin.test.tsx (hunk 10 above)
Comment (KishoreKicha14) [reply public/plugin.test.tsx]: reused Coremock

File: public/plugin.test.tsx (hunk 9 above)
Comment (KishoreKicha14) [reply public/plugin.test.tsx]: reused Coremock

File: common/utils/QueryUtils.test.ts [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+
+    const result = await retrieveQueryById(mockCore, dataSourceId, start, end, id);
+    expect(result).toEqual({
+      id: '1e5fde5b-c85f-419e-b8b6-3f43b3da4d59',
+      query: JSON.stringify({ match: { user_action: 'login_attempt' } }),
+    });
+  });
+
+  it('should return null if no top queries are found', async () => {
+    mockHttpGet.mockResolvedValue({ response: { top_queries: [] } });
```
[the remark this answers is not in the document]
Comment (KishoreKicha14) [reply common/utils/QueryUtils.test.ts]: added

**Comment by KishoreKicha14** [comment]
> We can also add unit tests for `MetricUtils` and `DateUtils`. Can do it as a followup?

Sure

File: public/plugin.test.tsx (hunk 12 above)
Comment (KishoreKicha14) [reply public/plugin.test.tsx]: Modified

**Review by deshsidd — APPROVED** [review]

**Comment by deshsidd** [comment]
@ansjcy Please take a look. LGTM. Thanks for address all the comments @KishoreKicha14 !

**Review by ansjcy — APPROVED** [review]
LGTM, thanks for the change!
